src/envoy/utils/utils.cc

```diff
@@ -118,6 +120,25 @@ bool GetPrincipal(const Network::Connection* connection, bool peer,
   return false;
 }
 
+bool GetSourceNamespace(const std::string& principal,
+                        std::string* source_namespace) {
+  if (source_namespace) {
```

Comment:
Must "ns/" be the start of the principal string? From the attribute builder test, looks like it not. However, if so, your extraction may have problem with principal like "sa/test_user_ns/ns/namespace_foo/bar"

Reply:
Good point Diem. The current code ([security/pkg/pki/ca/controller/secret.go](https://github.com/istio/istio/blob/85deb93bb837be26953b4bd3c44657dd488a4d93/security/pkg/pki/ca/controller/secret.go#L301)) generates it by putting the namespace in the beginning of the string, so this should be fine.

This should be stable for now, I also made the code a little more flexible, so it should still work for any small and reasonable changes to the format.

But there is a bigger potential issue, the format could be different in other environments. For example the managed CA might generate it in a different way. In this case we have to update our code anyway.

@wattli @myidpt, @wenchenglu  WDYT?

Reply:
I would suggest checking the prefix of the original SAN (i.e spiffe://) then switch to the proper extraction method. You may bundle all these into the GetPrincipal function, as that prefix was stripped off in that function.

Regarding to the code, if you expect "ns/" must be the start of the string, then use StartWith (or equivalent) instead.

Reply:
I feel it's better to not to couple these logic into one function, the `GetPrincipal()` is the base function that extracts the whole principal string, then we could use `GetSourceName()` or other similar functions (e.g. `GetServiceAccount()`) based on the principal returned from `GetPrincipal()`.

Also I refactored the code so it doesn't enforce the order of the "ns/".

Reply:
It depends. If you can always extract namespace from principal with fix schema, then this could be ok. However, if it's depends on the type of identity or other certificate information, then it's actually may be better to use the whole certificate. 
Bundle with extract Principal can also make logic consistent in they way we require identity must follow certain format (i.e spiffe:// prefix, contains of "ns/" etc.

Anyway, it's ok to do this later.